File: ivf-client/src/app/core/services/fingerprint/fingerprint-hub.service.ts
import { Injectable, inject, signal } from '@angular/core';
import * as signalR from '@microsoft/signalr';
import { environment } from '../../../../environments/environment';
import { Subject } from 'rxjs';

export interface CaptureRequest {
    patientId: string;
    fingerType: string;
    requestedBy: string;
    requestedAt: Date;
}

export interface CaptureResult {
    patientId: string;
    success: boolean;
    templateData?: string;   // Base64 encoded fingerprint template
    imageData?: string;      // Base64 encoded fingerprint image (JPEG)
    fingerType?: string;
    quality?: string;
    errorMessage?: string;
    capturedAt: Date;
}

export interface CaptureStatus {
    patientId: string;
    status: string;   // "Starting", "WaitingForFinger", "Capturing", "Processing", "Completed", "Failed"
    message?: string;
    timestamp: Date;
}

export interface SampleQuality {
    patientId: string;
    quality: string;   // "Good", "Poor", "TooWet", "TooDry"
    imagePreview?: string;
    timestamp: Date;
}

export interface EnrollmentProgress {
    patientId: string;
    samplesCollected: number;
    samplesNeeded: number;
    timestamp: Date;
}

export interface VerificationResult {
    patientId: string;
    success: boolean;
    fingerType?: string;
    errorMessage?: string;
    verifiedAt: Date;
}

@Injectable({ providedIn: 'root' })
export class FingerprintHubService {
    private hubConnection: signalR.HubConnection | null = null;
    private readonly hubUrl = `${environment.apiUrl.replace('/api', '')}/hubs/fingerprint`;

    // Reactive state
    isConnected = signal(false);
    connectionError = signal<string | null>(null);
    currentPatientId = signal<string | null>(null);

    // Events as Subjects for components to subscribe
    captureRequested$ = new Subject<CaptureRequest>();
    captureResult$ = new Subject<CaptureResult>();
    captureStatus$ = new Subject<CaptureStatus>();
    sampleQuality$ = new Subject<SampleQuality>();
    enrollmentProgress$ = new Subject<EnrollmentProgress>();
    captureCancelled$ = new Subject<string>();
    verificationResult$ = new Subject<VerificationResult>();

    /**
     * Connect to the FingerprintHub with JWT authentication
     */
    async connect(accessToken: string): Promise<boolean> {
        if (this.hubConnection?.state === signalR.HubConnectionState.Connected) {
            console.log('[FingerprintHub] Already connected');
            return true;
        }

        try {
            this.hubConnection = new signalR.HubConnectionBuilder()
                .withUrl(this.hubUrl, {
                    accessTokenFactory: () => accessToken
                })
                .withAutomaticReconnect([0, 2000, 5000, 10000, 30000])
                .configureLogging(signalR.LogLevel.Information)
                .build();

            this.setupEventHandlers();

            await this.hubConnection.start();
            this.isConnected.set(true);
            this.connectionError.set(null);
            console.log('[FingerprintHub] Connected successfully');
            return true;
        } catch (error) {
            console.error('[FingerprintHub] Connection failed:', error);
            this.isConnected.set(false);
            this.connectionError.set(error instanceof Error ? error.message : 'Connection failed');
            return false;
        }
    }

    /**
     * Disconnect from the hub
     */
    async disconnect(): Promise<void> {
        if (this.hubConnection) {
            await this.hubConnection.stop();
            this.isConnected.set(false);
            this.currentPatientId.set(null);
            console.log('[FingerprintHub] Disconnected');
        }
    }

    /**
     * Join a patient-specific capture group
     */
    async joinPatientCapture(patientId: string): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            throw new Error('Hub not connected');
        }

        await this.hubConnection.invoke('JoinPatientCapture', patientId);
        this.currentPatientId.set(patientId);
        console.log('[FingerprintHub] Joined patient capture:', patientId);
    }

    /**
     * Leave a patient-specific capture group
     */
    async leavePatientCapture(patientId: string): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            return;
        }

        await this.hubConnection.invoke('LeavePatientCapture', patientId);
        if (this.currentPatientId() === patientId) {
            this.currentPatientId.set(null);
        }
        console.log('[FingerprintHub] Left patient capture:', patientId);
    }

    /**
     * Request fingerprint capture for a patient
     * WinForms app will receive this and open capture form
     */
    async requestCapture(patientId: string, fingerType: string): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            throw new Error('Hub not connected');
        }

        await this.hubConnection.invoke('RequestCapture', patientId, fingerType);
        console.log('[FingerprintHub] Capture requested for:', patientId, fingerType);
    }

    /**
     * Request fingerprint verification for a patient
     */
    async requestVerification(patientId: string): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            throw new Error('Hub not connected');
        }

        await this.hubConnection.invoke('RequestVerification', patientId);
        console.log('[FingerprintHub] Verification requested for:', patientId);
    }

    /**
     * Cancel an ongoing capture request
     */
    async cancelCapture(patientId: string): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            return;
        }

        await this.hubConnection.invoke('CancelCapture', patientId);
        console.log('[FingerprintHub] Capture cancelled for:', patientId);
    }

    /**
     * Send capture result (used by WinForms adapter if needed)
     */
    async sendCaptureResult(result: CaptureResult): Promise<void> {
        if (!this.hubConnection || this.hubConnection.state !== signalR.HubConnectionState.Connected) {
            throw new Error('Hub not connected');
        }

        await this.hubConnection.invoke('SendCaptureResult', result);
    }

    private setupEventHandlers(): void {
        if (!this.hubConnection) return;

        // Reconnection handling
        this.hubConnection.onreconnecting((error) => {
            console.log('[FingerprintHub] Reconnecting...', error);
            this.isConnected.set(false);
        });

        this.hubConnection.onreconnected((connectionId) => {
            console.log('[FingerprintHub] Reconnected:', connectionId);
            this.isConnected.set(true);

            // Rejoin patient group if we were in one
            const patientId = this.currentPatientId();
            if (patientId) {
                this.joinPatientCapture(patientId).catch(console.error);
            }
        });

        this.hubConnection.onclose((error) => {
            console.log('[FingerprintHub] Connection closed:', error);
            this.isConnected.set(false);
        });

        // Listen for events from hub
        this.hubConnection.on('JoinedCapture', (patientId: string) => {
            console.log('[FingerprintHub] Joined capture confirmed:', patientId);
        });

        this.hubConnection.on('CaptureRequested', (request: CaptureRequest) => {
            console.log('[FingerprintHub] Capture requested:', request);
            this.captureRequested$.next(request);
        });

        this.hubConnection.on('CaptureResult', (result: CaptureResult) => {
            console.log('[FingerprintHub] Capture result:', result);
            this.captureResult$.next(result);
        });

        this.hubConnection.on('CaptureStatus', (status: CaptureStatus) => {
            console.log('[FingerprintHub] Capture status:', status);
            this.captureStatus$.next(status);
        });

        this.hubConnection.on('SampleQuality', (quality: SampleQuality) => {
            console.log('[FingerprintHub] Sample quality:', quality);
            this.sampleQuality$.next(quality);
        });

        this.hubConnection.on('EnrollmentProgress', (progress: EnrollmentProgress) => {
            console.log('[FingerprintHub] Enrollment progress:', progress);
            this.enrollmentProgress$.next(progress);
        });

        this.hubConnection.on('CaptureCancelled', (patientId: string) => {
            console.log('[FingerprintHub] Capture cancelled:', patientId);
            this.captureCancelled$.next(patientId);
        });

        this.hubConnection.on('VerificationResult', (result: VerificationResult) => {
            console.log('[FingerprintHub] Verification result:', result);
            this.verificationResult$.next(result);
        });
    }
}

File: src/IVF.FingerprintClient/Services/FingerprintHubService.cs
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.DependencyInjection;

namespace IVF.FingerprintClient.Services;

/// <summary>
/// Service for connecting to the IVF API FingerprintHub via SignalR.
/// </summary>
public class FingerprintHubService : IAsyncDisposable
{
    private HubConnection? _hubConnection;
    private readonly string _hubUrl;
    private string? _currentPatientId;

    public event EventHandler<CaptureRequestEventArgs>? CaptureRequested;
    public event EventHandler<VerificationRequestEventArgs>? VerificationRequested;
    public event EventHandler? CaptureCancelled;
    public event EventHandler<string>? StatusChanged;
    public event EventHandler<bool>? ConnectionChanged;

    public bool IsConnected => _hubConnection?.State == HubConnectionState.Connected;
    public string? CurrentPatientId => _currentPatientId;

    /// <summary>
    /// Send verification result to the Angular UI.
    /// </summary>
    public async Task SendVerificationResultAsync(VerificationResultDto result)
    {
        if (_hubConnection is null || !IsConnected)
        {
            throw new InvalidOperationException("Not connected to hub");
        }

        try
        {
            StatusChanged?.Invoke(this, $"Gửi verification result cho patient: {result.PatientId}");
            await _hubConnection.SendAsync("SendVerificationResult", result);
            StatusChanged?.Invoke(this, result.Success ? "Đã gửi verification thành công" : "Đã gửi verification thất bại");
        }
        catch (Exception ex)
        {
            StatusChanged?.Invoke(this, $"Lỗi gửi verification result: {ex.Message}");
            throw;
        }
    }

    public FingerprintHubService(string hubUrl)
    {
        _hubUrl = hubUrl;
    }

    /// <summary>
    /// Connect to the FingerprintHub with API key for desktop client authentication.
    /// </summary>
    public async Task<bool> ConnectAsync(string? apiKey = null)
    {
        try
        {
            var builder = new HubConnectionBuilder()
                .WithUrl(_hubUrl, options =>
                {
                    // Add API key to query string for desktop client authentication
                    if (!string.IsNullOrEmpty(apiKey))
                    {
                        options.Headers.Add("X-API-Key", apiKey);
                        // Also add to query string as backup
                        var uriBuilder = new UriBuilder(_hubUrl);
                        var query = System.Web.HttpUtility.ParseQueryString(uriBuilder.Query);
                        query["apiKey"] = apiKey;
                        uriBuilder.Query = query.ToString();
                        options.Url = uriBuilder.Uri;
                    }
                })
                .WithAutomaticReconnect(new[] { TimeSpan.Zero, TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(10) })
                .AddJsonProtocol(options =>
                {
                    options.PayloadSerializerOptions.PropertyNameCaseInsensitive = true;
                });

            _hubConnection = builder.Build();

            SetupEventHandlers();

            await _hubConnection.StartAsync();
            StatusChanged?.Invoke(this, "Đã kết nối đến server");
            ConnectionChanged?.Invoke(this, true);
            return true;
        }
        catch (Exception ex)
        {
            StatusChanged?.Invoke(this, $"Lỗi kết nối: {ex.Message}");
            ConnectionChanged?.Invoke(this, false);
            return false;
        }
    }

    /// <summary>
    /// Disconnect from the hub.
    /// </summary>
    public async Task DisconnectAsync()
    {
        if (_hubConnection != null)
        {
            await _hubConnection.StopAsync();
            StatusChanged?.Invoke(this, "Đã ngắt kết nối");
            ConnectionChanged?.Invoke(this, false);
        }
    }

    /// <summary>
    /// Join a patient capture group to listen for capture requests.
    /// </summary>
    public async Task JoinPatientCaptureAsync(string patientId)
    {
        if (_hubConnection is null || !IsConnected)
        {
            throw new InvalidOperationException("Not connected to hub");
        }

        await _hubConnection.InvokeAsync("JoinPatientCapture", patientId);
        _currentPatientId = patientId;
        StatusChanged?.Invoke(this, $"Đã tham gia nhóm bệnh nhân: {patientId}");
    }

    /// <summary>
    /// Leave a patient capture group.
    /// </summary>
    public async Task LeavePatientCaptureAsync(string patientId)
    {
        if (_hubConnection is null || !IsConnected) return;

        await _hubConnection.InvokeAsync("LeavePatientCapture", patientId);
        if (_currentPatientId == patientId)
        {
            _currentPatientId = null;
        }
    }

    /// <summary>
    /// Send capture result to the Angular UI.
    /// </summary>
    public async Task SendCaptureResultAsync(CaptureResultDto result)
    {
        if (_hubConnection is null || !IsConnected)
        {
            throw new InvalidOperationException("Not connected to hub");
        }

        try
        {
            StatusChanged?.Invoke(this, $"Gửi result cho patient: {result.PatientId}");
            // Use SendAsync (fire-and-forget) instead of InvokeAsync since hub method returns void
            await _hubConnection.SendAsync("SendCaptureResult", result);
            StatusChanged?.Invoke(this, result.Success ? "Đã gửi kết quả thành công" : "Đã gửi kết quả thất bại");
        }
        catch (Exception ex)
        {
            StatusChanged?.Invoke(this, $"Lỗi gửi kết quả: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Send capture status update.
    /// </summary>
    public async Task SendCaptureStatusAsync(string patientId, string status, string? message = null)
    {
        if (_hubConnection is null || !IsConnected) return;

        await _hubConnection.SendAsync("SendCaptureStatus", patientId, status, message);
    }

    /// <summary>
    /// Send sample quality feedback.
    /// </summary>
    public async Task SendSampleQualityAsync(string patientId, string quality, string? imageData = null)
    {
        if (_hubConnection is null || !IsConnected) return;

        await _hubConnection.SendAsync("SendSampleQuality", patientId, quality, imageData);
    }

    /// <summary>
    /// Send enrollment progress.
    /// </summary>
    public async Task SendEnrollmentProgressAsync(string patientId, int samplesCollected, int samplesNeeded)
    {
        if (_hubConnection is null || !IsConnected) return;

        await _hubConnection.SendAsync("SendEnrollmentProgress", patientId, samplesCollected, samplesNeeded);
    }

    private void SetupEventHandlers()
    {
        if (_hubConnection is null) return;

        _hubConnection.Reconnecting += error =>
        {
            StatusChanged?.Invoke(this, "Đang kết nối lại...");
            ConnectionChanged?.Invoke(this, false);
            return Task.CompletedTask;
        };

        _hubConnection.Reconnected += async connectionId =>
        {
            StatusChanged?.Invoke(this, "Đã kết nối lại");
            ConnectionChanged?.Invoke(this, true);
            
            // Auto-rejoin patient group if we were in one
            if (!string.IsNullOrEmpty(_currentPatientId))
            {
                try
                {
                    StatusChanged?.Invoke(this, $"Đang rejoin group cho patient: {_currentPatientId}");
                    await JoinPatientCaptureAsync(_currentPatientId);
                    StatusChanged?.Invoke(this, "Đã rejoin group thành công!");
                }
                catch (Exception ex)
                {
                    StatusChanged?.Invoke(this, $"Lỗi rejoin group: {ex.Message}");
                }
            }
        };

        _hubConnection.Closed += error =>
        {
            StatusChanged?.Invoke(this, "Đã mất kết nối");
            ConnectionChanged?.Invoke(this, false);
            return Task.CompletedTask;
        };

        // Listen for capture requests from Angular UI
        _hubConnection.On<CaptureRequestDto>("CaptureRequested", async request =>
        {
            _currentPatientId = request.PatientId;
            StatusChanged?.Invoke(this, $"Nhận request cho patient: {request.PatientId}");
            
            // Join the patient group so we can send results back
            try
            {
                StatusChanged?.Invoke(this, $"Đang join group: fingerprint_{request.PatientId}");
                await JoinPatientCaptureAsync(request.PatientId);
                StatusChanged?.Invoke(this, $"Đã join group thành công!");
            }
            catch (Exception ex)
            {
                StatusChanged?.Invoke(this, $"Lỗi join group: {ex.Message}");
            }
            
            CaptureRequested?.Invoke(this, new CaptureRequestEventArgs(request));
        });

            // Listen for verification requests
            _hubConnection.On<VerificationRequestDto>("VerificationRequested", async dto =>
            {
                var request = dto.Request;
                _currentPatientId = request.PatientId;
                StatusChanged?.Invoke(this, $"Nhận verification request cho patient: {request.PatientId}");
                
                try
                {
                    StatusChanged?.Invoke(this, $"Đang join group cho verification: fingerprint_{request.PatientId}");
                    await JoinPatientCaptureAsync(request.PatientId);
                }
                catch (Exception ex)
                {
                    StatusChanged?.Invoke(this, $"Lỗi join group: {ex.Message}");
                }
                
                VerificationRequested?.Invoke(this, new VerificationRequestEventArgs(dto));
            });
// ... (lines 271-351 unchanged)



        // Listen for capture cancellation
        _hubConnection.On<string>("CaptureCancelled", patientId =>
        {
            if (patientId == _currentPatientId)
            {
                CaptureCancelled?.Invoke(this, EventArgs.Empty);
            }
        });

        // Confirmation of joining
        _hubConnection.On<string>("JoinedCapture", patientId =>
        {
            StatusChanged?.Invoke(this, $"Đã xác nhận tham gia: {patientId}");
        });
    }

    public async ValueTask DisposeAsync()
    {
        if (_hubConnection != null)
        {
            await _hubConnection.DisposeAsync();
        }
    }
}

// DTOs matching the server-side definitions
public record CaptureRequestDto
{
    public string PatientId { get; init; } = string.Empty;
    public string FingerType { get; init; } = string.Empty;
    public string RequestedBy { get; init; } = string.Empty;
    public DateTime RequestedAt { get; init; }
}

public record CaptureResultDto
{
    public string PatientId { get; init; } = string.Empty;
    public bool Success { get; init; }
    public string? TemplateData { get; init; }  // Base64 encoded fingerprint template
    public string? ImageData { get; init; }     // Base64 encoded fingerprint image (JPEG)
    public string? FingerType { get; init; }
    public string? Quality { get; init; }
    public string? ErrorMessage { get; init; }
    public DateTime CapturedAt { get; init; }
}

public class CaptureRequestEventArgs : EventArgs
{
    public CaptureRequestDto Request { get; }

    public CaptureRequestEventArgs(CaptureRequestDto request)
    {
        Request = request;
    }
}

// DTOs matching the server-side definitions
public record VerificationRequestDto
{
    public VerificationRequestData Request { get; init; } = new();
}

public record VerificationRequestData
{
    public string PatientId { get; init; } = string.Empty;
    public string RequestedBy { get; init; } = string.Empty;
    public DateTime RequestedAt { get; init; }
    public List<FingerprintTemplateDto> Templates { get; init; } = new();
}

public record FingerprintTemplateDto
{
    public string FingerType { get; init; } = string.Empty;
    public string TemplateData { get; init; } = string.Empty;
}

public record VerificationResultDto
{
    public string PatientId { get; init; } = string.Empty;
    public bool Success { get; init; }
    public string? FingerType { get; init; }
    public string? ErrorMessage { get; init; }
    public DateTime VerifiedAt { get; init; }
}

public class VerificationRequestEventArgs : EventArgs
{
    public VerificationRequestData Request { get; }
    public VerificationRequestEventArgs(VerificationRequestDto dto) => Request = dto.Request;
}
